test/format/thread_pause.c

```diff
@@ -0,0 +1,129 @@
+/*-
+ * Public Domain 2014-present MongoDB, Inc.
+ * Public Domain 2008-2014 WiredTiger, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "format.h"
+
+/*
+ * FIXME-thread-pause There's a lot of white boxing here where we reach into the SESSION_IMPL and
+ * CONNECTION_IMPL structs. The alternative is a handful of new API functions but I don't like that
+ * idea.
+ */
+
+/*
+ * thread_pause --
+ *     Randomly pick active threads as tracked in the connections thread_registry and pause them for
+ *     a brief period of time. The OS won't let me play with the scheduler so we use pthread signals
+ *     instead.
+ */
+WT_THREAD_RET
+thread_pause(void *arg)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_SESSION_IMPL *session;
+    int ret;
+
+    WT_RAND_STATE rnd;
+    uint32_t rand_index;
+    wt_thread_t *rand_thread;
+    char thread_name[WT_THREAD_NAME_MAX_LEN];
+
+    session = (WT_SESSION_IMPL *)arg;
+    conn = S2C(session);
+
+    __wt_random_init_seed(NULL, &rnd);
+
+    while (!g.workers_finished) {
+        __wt_readlock(session, &conn->internal_thread_registry_lock);
+
+        /* The thread registry array can have holes. Keep trying until we find something */
+        rand_thread = NULL;
+        while (rand_thread == NULL && !g.workers_finished) {
+            rand_index = __wt_random(&rnd) % conn->internal_thread_registry_size;
+            rand_thread = conn->internal_thread_registry[rand_index];
+        }
```

Comment:
If there are a large number of holes (maybe a lot of threads died except for the "last" one), this might miss for a while, all while holding a lock.  Maybe pick one random number and scan (circularly) from that point looking for non-NULL.  If you went completely around the circle you'd know that something went wrong.